libtiledbvcf/src/read/reader.cc

```diff
@@ -894,6 +1186,258 @@ std::vector<SampleAndId> Reader::prepare_sample_names() const {
+      auto it = md.sample_ids.find(s);
+      if (it == md.sample_ids.end())
+        throw std::runtime_error(
+            "Error preparing sample list for export; sample '" + s +
+            "' has not been registered.");
+      result.push_back({.sample_name = s, .sample_id = it->second});
+    }
+  }
+
+  return result;
+}
+
+void Reader::prepare_regions_v4(
+    std::vector<Region>* regions,
+    std::unordered_map<std::string, std::vector<size_t>>*
+        regions_index_per_contig,
+    std::vector<std::pair<std::string, std::vector<QueryRegion>>>*
+        query_regions) const {
+  const uint32_t g = dataset_->metadata().anchor_gap;
+  // Use a linked list for pre-partition regions to allow for parallel parsing
+  // of BED file
+  std::list<Region> pre_partition_regions_list;
+
+  // Manually-specified regions (-r) are 1-indexed and inclusive
+  for (const std::string& r : params_.regions)
+    pre_partition_regions_list.emplace_back(
+        r, Region::Type::OneIndexedInclusive);
+
+  // Add BED file regions, if specified.
+  if (!params_.regions_file_uri.empty()) {
+    auto start_bed_file_parse = std::chrono::steady_clock::now();
+    Region::parse_bed_file_htslib(
+        params_.regions_file_uri, &pre_partition_regions_list);
+    if (params_.verbose) {
+      auto old_locale = std::cout.getloc();
+      utils::enable_pretty_print_numbers(std::cout);
+      std::cout << "Parsed bed file into " << pre_partition_regions_list.size()
+                << " regions in "
+                << utils::chrono_duration(start_bed_file_parse) << " seconds."
+                << std::endl;
+      std::cout.imbue(old_locale);
+    }
+  }
+
+  // No specified regions means all regions.
+  if (pre_partition_regions_list.empty())
+    pre_partition_regions_list = dataset_->all_contigs_list();
+
+  Array array = Array(*ctx_, dataset_->data_uri(), TILEDB_READ);
+  std::pair<uint32_t, uint32_t> regionNonEmptyDomain =
+      array.non_empty_domain<uint32_t>(1);
+  std::vector<Region> filtered_regions;
+  // Loop through all contigs to query and pre-filter to ones which fall
+  // inside the nonEmptyDomain This will balance the partitioning better my
+  // removing empty regions
+  for (auto& r : pre_partition_regions_list) {
+    uint32_t contig_offset;
+    try {
+      contig_offset = dataset_->metadata().contig_offsets.at(r.seq_name);
+    } catch (const std::out_of_range&) {
+      throw std::runtime_error(
+          "Error preparing regions for export; no contig named '" + r.seq_name +
+          "' in dataset.");
+    }
+
+    r.seq_offset = contig_offset;
+    const uint32_t reg_min = r.min;
+    const uint32_t reg_max = r.max;
+
+    // Widen the query region by the anchor gap value, avoiding overflow.
+    uint64_t widened_reg_min = g > reg_min ? 0 : reg_min - g;
+    if (widened_reg_min <= regionNonEmptyDomain.second &&
+        reg_max >= regionNonEmptyDomain.first) {
+      filtered_regions.emplace_back(r);
```

Comment:
if `Region` has a move constructor, it looks like it makes sense to use that here since this is the last use of the element on `pre_partition_regions_list`:
```
filtered_regions.emplace_back(std::move(r));
```